crates/sui-indexer/src/models/watermarks.rs

```diff
@@ -0,0 +1,176 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::schema::watermarks::{self};
+use diesel::prelude::*;
+
+/// Represents a row in the `watermarks` table.
+#[derive(Queryable, Insertable, Default)]
+#[diesel(table_name = watermarks, primary_key(entity))]
+pub struct StoredWatermark {
+    /// The table name of group of tables governed by this watermark, i.e `epochs`, `checkpoints`,
+    /// `transactions`.
+    pub entity: String,
+    /// Upper bound epoch range to enable per-entity epoch-level retention policy. Committer
+    /// advances this along with `high`.
+    pub epoch_hi: i64,
+    /// Lower bound epoch range to enable per-entity epoch-level retention policy. Pruner advances
+    /// this.
+    pub epoch_lo: i64,
+    /// The inclusive high watermark that the committer advances.
+    pub hi: i64,
```

Comment:
when `hi` means transaction seq number, and when the ingestion task crashes and resumes, it needs a checkpoint seq number instead to fetch the corresponding checkpoint data, so wondering what benefits do transaction seq number brings us, shall we all use checkpoint seq instead?

Reply:
The lo and hi tracks the progress (ingestion and pruning) of each atomic entity, and I make the assumption that we might eventually get to a point where transactions, events, etc can be ingested and pruned independently. If that is the case, then if checkpoint sequence number is the smallest unit tracked on the watermark, then we'd need to consult another table to understand the corresponding range 

Alternatively, we can directly track each entity's range to reduce the number of round trips we need to make

Since ingestion task operates on the checkpoint level, I think we can just add a checkpoint_lo to each entry so it knows where to pick up?

Thus, epoch_lo and hi for pruner to determine when to prune (per table retention policy)

Checkpoint_lo (hi probably isn't needed) for ingestion task to know where to restart

Lo and hi for reader

Reply:
> I think we can just add a checkpoint_lo to each entry so it knows where to pick up

yes sounds good, although the naming gets a bit puzzling now, if we stick to the idea that ingestion ~= hi, pruning ~= hi, should this be hi instead? to be exact, I think here the watermark means the checkpoint number where all previous checkpoints have been totally ingested.

Reply:
Haha good call
Perhaps I gotta think of a better naming convention but yes you're right

Reply:
It took me a while to figure out how exactly the ingestion task knows where to resume today. It seems to be in the `ShimIndexerProgressStore` implements `ProgressStore`. For each task it stores the checkpoint sequence number it processed. in `start_writer_with_config`, we create two executors, `primary` and `object_snapshot`

I can introduce `checkpoint_hi` and have `committer` write to it, but do you think I can hook this into the data ingestion stuff in a follow-up PR?

Reply:
yeah the split makes sense!